### ISSUE TITLE
bcachefs: Fails with `device = "UUID=..."`

### ISSUE THREAD

**Issue by ElvishJerricco** [NixOS/nixpkgs#317901]
When you specify the device with `UUID=...`, the `unlock-bcachefs-` service [attempts to wait on a device](https://github.com/NixOS/nixpkgs/blob/6e5c9a8064826812096631ff364eb3622fa7efa6/nixos/modules/tasks/filesystems/bcachefs.nix#L83-L85) called `UUID=...`. This doesn't work, because systemd does not translate this into a device path; it just expects a literal file named `UUID=...`. This has been fixed *specifically* for the `What=` directive in mount units, but this fix does not apply to device unit names.

The result is that you end up waiting on a device that will never exist to run a service that you might not even need if you're not using encryption (which is normally fine because of the `ExecCondition`).

We've really gotta rethink everything about how we're mounting bcachefs file systems, especially given the problems with multi-device ones.

**Comment by JohnRTitor** [comment]
A workaround here: https://github.com/NixOS/nixpkgs/pull/316695#issuecomment-2143984073

Apparently this is a systemd bug: https://github.com/NixOS/nixpkgs/pull/316695#issuecomment-2144630389 (Upstream issue: https://github.com/systemd/systemd/issues/28989)

CC: @lheckemann 

I feel like a lot of issues like https://github.com/NixOS/nixpkgs/issues/316396 these keep popping up due to slow moving systemd and incompatibility with our current mounting methods.

**Comment by ElvishJerricco** [comment]
That workaround does work, because it switches from `UUID=` to an absolute device path. However that systemd bug is not at all relevant in this issue. It's entirely our own fault with how we generate our `unlock-bcachefs-*` services.

**Comment by ElvishJerricco** [comment]
I may have gone a little overboard and created this thing to solve this: https://github.com/ElvishJerricco/bcachefs-fstab-generator

TL;DR: It's a systemd generator that parses `fstab` to generate the unlock service, rather than having Nix expressions create services for it. It's not fundamentally all that different from what NixOS is doing, except it can also do `gpt-auto-root`. But it can also parse `UUID=...` style devices correctly too :P

**Comment by nyabinary** [comment]
Is this planned to be solved anytime soon?

**Comment by ElvishJerricco** [comment]
To be clear, this issue is trivial to workaround by replacing `device = "UUID=...";` with `device = "/dev/disk/by-uuid/...";`.

**Comment by ElvishJerricco** [comment]
Finally created a PR to fix this. Please review. #419830
